ui/src/app/utils/authentication.tsx

```diff
@@ -0,0 +1,114 @@
+import {authStore, serverConfigStore, useStore} from 'app/utils/stores';
+import {environment} from 'environments/environment';
+import {ConfigResponse} from 'generated/fetch';
+import {useEffect} from 'react';
+import {AnalyticsTracker, setLoggedInState} from './analytics';
+import {LOCAL_STORAGE_KEY_TEST_ACCESS_TOKEN} from './cookies';
+import {signInStore} from './navigation';
+
+declare const gapi: any;
+
+// for e2e tests: provide your own oauth token to obviate Google's oauth UI
+// flow, thereby avoiding inevitable challenges as Google identifies Puppeteer
+// as non-human.
+declare global {
+  interface Window { setTestAccessTokenOverride: (token: string) => void; }
+}
+
+const makeAuth2 = (config: ConfigResponse): Promise<any> => {
+  return new Promise((resolve) => {
+    gapi.load('auth2', () => {
+      console.log(config);
+      gapi.auth2.init({
+        client_id: environment.clientId,
+        hosted_domain: config.gsuiteDomain,
+        scope: 'https://www.googleapis.com/auth/plus.login openid profile'
+            + (config.enableBillingUpgrade ? ' https://www.googleapis.com/auth/cloud-billing' : '')
+      }).then(() => {
+        console.log('Setting auth loaded, isSignedIn');
+        authStore.set({
+          ...authStore.get(),
+          authLoaded: true,
+          isSignedIn: gapi.auth2.getAuthInstance().isSignedIn.get(),
+          enableInactivityTimeout: false
+        });
+
+        console.log('Setting up gapi auth listener; isSignedIn: ', gapi.auth2.getAuthInstance().isSignedIn.get());
+        gapi.auth2.getAuthInstance().isSignedIn.listen((nextIsSignedIn: boolean) => {
+          authStore.set({...authStore.get(), isSignedIn: nextIsSignedIn});
```

Comment:
I just realized what has been confusing me so much about this authentication logic and also how we use our stores in general.

Because we sometimes use stores to set the value for what we want the application to be (aka actions in redux) our stores can be in a state where the value is not exactly true and our logic needs to deduce if we need to make the application state correct by acting on a new store value.

Applied to this case, we need to set up some code elsewhere in the application that listens for `isSignedIn` being set to `false` and then run the sign out flow.

I'm thinking it might be easier if we just run the sign out code here and then set the store value to reflect that.

Reply:
To make sure I'm understanding what you're saying correctly - are you suggesting a `useEffect` here depending on `isSignedIn` to sign out if you're not actually signed in? Like splitting up the second `useEffect` in `useAuthentication`? Or are you suggesting moving away from the Redux-actions-style approach?

Reply:
I think I get it now. Previous consumers of `signOut` should instead set the store value. They'll have to go back to handling their own navigate-after-signout, but that's fine, there's not many of them.